hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -3040,45 +3024,61 @@ public TermIndex installSnapshot(String leaderId) {
     } catch (Exception e) {
       LOG.error("Failed to stop/ pause the services. Cannot proceed with " +
           "installing the new checkpoint.", e);
+
+      // During stopServices, if KeyManager was stopped successfully and
+      // OMMetadataManager stop failed, we should restart the KeyManager.
+      keyManager.start(configuration);
+
       return null;
     }
 
-    //TODO: un-pause SM if any failures and retry?
+    File dbBackup;
+    TermIndex termIndex = omRatisServer.getLastAppliedTermIndex();
+    long currentTerm = termIndex.getTerm();
+    long lastAppliedIndex = termIndex.getIndex();
+    boolean loadSuccess = false;
 
-    long lastAppliedIndex = omRatisServer.getLastAppliedTermIndex().getIndex();
+    try {
+      // Check if current applied log index is smaller than the downloaded
+      // checkpoint transaction index. If yes, proceed by stopping the ratis
+      // server so that the OM state can be re-initialized. If no then do not
+      // proceed with installSnapshot.
+      boolean canProceed = OzoneManagerRatisUtils.verifyTransactionInfo(
+          omTransactionInfo, lastAppliedIndex, leaderId, newDBLocation);
+      if (!canProceed) {
+        return null;
+      }
 
-    boolean canProceed =
-        OzoneManagerRatisUtils.verifyTransactionInfo(omTransactionInfo,
-        lastAppliedIndex, leaderId, newDBlocation);
+      try {
+        dbBackup = replaceOMDBWithCheckpoint(lastAppliedIndex, oldDBLocation,
+            newDBLocation);
+      } catch (Exception e) {
+        LOG.error("OM DB checkpoint replacement with new downloaded " +
+            "checkpoint failed.", e);
+        return null;
+      }
 
-    // If downloaded DB has transaction info less than current one, return.
-    if (!canProceed) {
-      return null;
+      loadSuccess = true;
+    } finally {
+      if (!loadSuccess) {
+        omRatisServer.getOmStateMachine().unpause(lastAppliedIndex,
+            currentTerm);
+      }
     }
 
     long leaderIndex = omTransactionInfo.getTransactionIndex();
```

Comment:
Move these inside the try block, just after you invoke `replaceOMDBWithCheckpoint`. Else these variables should be left to values that reflect the old checkpoint.